KEP: 3rd party content in documentation (#1327)
kubernetes/enhancements: keps/sig-docs/20191020-doc-policies-for-third-party-content.md

---
title: doc-policies-for-third-party-content

> authors:

Reviewer: `@aimeeu` helped with the draft as well

reviewers:
  - TBD
approvers:
  - TBD
editor: TBD
creation-date: 2019-10-20
last-updated: 2019-10-20
status: provisional
---

# doc-policies-for-third-party-content

## Table of Contents

<!-- toc -->
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [User Stories](#user-stories)
    - [Story 1 (fictional)](#story-1-fictional)
    - [Story 2 (fictional)](#story-2-fictional)
    - [Story 3 (actual)](#story-3-actual)
    - [Story 4 (actual)](#story-4-actual)
    - [Story 5 (actual)](#story-5-actual)
  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Design Details](#design-details)
  - [Graduation Criteria](#graduation-criteria)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

*Note* that this KEP does not target any release; SIG Docs follows a continuous release process for website content.

## Summary

Feature docs are not a place for vendor pitches. Nevertheless, SIG Docs has seen pull requests that are more or less a request to place advertising content on the Kubernetes website.

> Feature docs also contain content hosted in multiple places (“dual-sourced content”). A good practice for code project docs is to host single-sourced content only, and to provide links to other providers’ single-sourced content.

Reviewer 2: not requesting clarification, but just wanted to mention the docker case here.
external websites already provide information on how to install docker, yet we have some further details in our guide:
- the newest docker version is usually not approved yet, thus our guides are usually trailing behind recommending an older / tested docker.
- distros or the docker project bundle a docker version that usually has `cgroupfs` as it's cgroup driver, we recommend changing that to `systemd`.

ref: https://kubernetes.io/docs/setup/production-environment/container-runtimes/#docker

Author: Docker (and other 3rd party content tied closely with Kubernetes) will need different treatment from, say, ACME Inc.'s Ingress controller. That's fine.

I'm not sure if there should be a defined list of projects with special treatments, or merely rough consensus and an appreciation that each case can have its own merits.

Reviewer 2: possibly list the exceptions as part of the KEP?

This KEP defines a policy on documentation content, so that authors can judge what is appropriate to propose and so that PR approvers can make consistent, fair decisions during the review process.

## Motivation

SIG Docs publishes Kubernetes documentation on kubernetes.io, in line with its [charter](https://github.com/kubernetes/community/blob/master/sig-docs/charter.md#scope), and also sets standards for website content. Prior to this KEP there were no clear guidelines or standards showing what documentation is acceptable.

The Kubernetes documentation is currently a mix of both 1) documentation describing the Kubernetes open source project; and 2) content describing how to install or use Kubernetes on several third party Kubernetes offerings.

Some third party content is necessary in order for Kubernetes to function. For example, Docker, networking policy (CNI plugins), Ingress controllers, and logging all require third party components.

> Pages like [Logging Using Elasticsearch and Kibana](https://kubernetes.io/docs/tasks/debug-application-cluster/logging-elasticsearch-kibana/) are highly specific to a third party offering and seem more like third party product documentation than Kubernetes open source documentation.

Author: AIUI, this paragraph aims to include:
- Docker
- CNI plugins
- Ingress controllers
- logging helpers

as places where 3rd party components are useful.

https://kubernetes.io/docs/tasks/debug-application-cluster/logging-elasticsearch-kibana/ seems like the sort of page where every vendor for a log service might reasonable want to put up a similar page about how to hook your cluster up to that logging service. This starts to look like vendor pitch territory.

@timothysc do you have some alternative wording in mind?

> Pages like [Logging Using Elasticsearch and Kibana](https://kubernetes.io/docs/tasks/debug-application-cluster/logging-elasticsearch-kibana/) are highly specific to a third party offering and seem more like third party product documentation than Kubernetes open source documentation.

Reviewer 3: > seems like the sort of page where every vendor for a log service might reasonable want to put up a similar page about how to hook your cluster up to that logging service. This starts to look like vendor pitch territory. 

This statement is a false choice ^.  If there is a community set of best practices that use OSS tools, and may or may not have a private backer, we should provide documentation on what we consider to be best practices using the public facing artifacts.  That said, we should remove private, proprietary integrations and point to external documentation.

> Pages like [Logging Using Elasticsearch and Kibana](https://kubernetes.io/docs/tasks/debug-application-cluster/logging-elasticsearch-kibana/) are highly specific to a third party offering and seem more like third party product documentation than Kubernetes open source documentation.

Reviewer 4: @timothysc Thanks for the review. I have questions.

One of the goals of this KEP is to achieve specific processes for how these things _actually get done_, not just by hand waving.

```
1. Who decides when to include third-party content?
1. What standard of quality and review must be met before docs include third-party content?
```

> We vet specific versions so there is a case by case basis for the changes.

When you say "we", who is that specifically? What is the specific vetting process?

> If there is a community set of best practices that use OSS tools, and may or may not have a private backer, **we should provide documentation on what we consider to be best practices** using the public facing artifacts. (emphasis added)

That is the motivation for this KEP. This is the place where we create specific practices for Kubernetes.

> This example is not a good one IMO. 

It's a real example in the docs that we have to deal with. How should we proceed?

> These are each OSS projects whose integration provides value to the community. 

Who specifically determines whether integrations provide value to the community and how?

> If they have a page related to kubernetes then it makes sense to not host. 

If projects have default permission to host on kubernetes.io, what is the incentive for projects to host their own documentation?

```
1. When should the Kubernetes documentation host third party documents that we are not the source/authority of?
1. How does the Kubernetes project handle third party content that is not kept up to date or hosts?
```

> However, in the early days of kubernetes, k8s was the plan of record.

Can you point to where that's documented? Is there a good reason for tradition to continue simply because [_that's the way we've always done it_](https://www.forbes.com/sites/forbeslacouncil/2019/01/28/the-most-dangerous-phrase-in-business-weve-always-done-it-this-way/#2c7c02140f7e)?

> Pages like [Logging Using Elasticsearch and Kibana](https://kubernetes.io/docs/tasks/debug-application-cluster/logging-elasticsearch-kibana/) are highly specific to a third party offering and seem more like third party product documentation than Kubernetes open source documentation.

Reviewer 4: The question of Elasticsearch is particularly compelling, given the [tension](https://news.ycombinator.com/item?id=19359602) surrounding it. I raise the example specifically because it isn't clear who vets these inclusions or how&mdash;and good docs need these processes to be clear and transparent, not just a matter of a single developer saying, "It's there because I said so."

> Pages like [Logging Using Elasticsearch and Kibana](https://kubernetes.io/docs/tasks/debug-application-cluster/logging-elasticsearch-kibana/) are highly specific to a third party offering and seem more like third party product documentation than Kubernetes open source documentation.

Reviewer 3: That seems odd.  I would expect the authoring sig would own and maintain.  In this case I would expect sig-apps to maintain this work.

The goal of this KEP is to reach and document a consensus on what types of third party content are appropriate for inclusion in Kubernetes documentation; standards for including third-party content; and to drive a consistent policy toward third party content inclusion.
This KEP focuses on the following issues related to third party content inclusion:

1. What third party content is appropriate for inclusion in the Kubernetes documentation?
1. Does third party content in sections such as [Getting Started](https://kubernetes.io/docs/setup/) in the docs provide sufficient value to the reader that they should remain?
1. Is there a list of content pages that are so focused on third party product usage that they should be removed or updated from the Kubernetes documentation?
1. When should the Kubernetes documentation host third party documents that we are not the source/authority of?
1. How does the Kubernetes project handle third party content that is not kept up to date or hosts?

> 1. Can feature owners flag when third party content is *required*, as opposed to preferable or common?

Reviewer 2: interested how this can get implemented mechanically.

Reviewer 4: @neolit123 Same!

Reviewer 5: Maybe through the process of this Kep we nail down a list?

1. Who decides when to include third-party content?
1. What standard of quality and review must be met before docs include third-party content?
1. To what extent should SIG Docs advocate for third-party content providers to host their own content, or decline to host third-party content altogether?

### Goals

- Policy on including content about third party products in Kubernetes web documentation.
- Published guidelines for contributors and reviewers that clarify requirements for third party content and allow documentation reviewers to reject advertising.

### Non-Goals

- Outright removal of all content relating to vendors and projects outside the CNCF ecosystem.

## Proposal

> TBD. Detailed policies will come out of the discussion around this KEP.

Reviewer 3: IMO just a simple decision flowchart should be sufficient to outline the process.

Author: Let's hope so!

Reviewer 4: > IMO just a simple decision flowchart should be sufficient to outline the process.

Agreed, that's an ideal outcome of this KEP.

Bob is a documentation approver. Bob explains that Alice’s proposed change does not meet community standards, because it is functionally an advertisement.

#### Story 2 (fictional)

Charlie uses Linux, specifically Ubuntu. Charlie notices that the page about installing `kubethingy` has instructions for installing `kubethingy` on Windows and on CentOS/RHEL but not on Ubuntu. Charlie reads the guidelines on content and sees that this kind of change is acceptable (Ubuntu is one of the most popular Linux distributions, and `kubethingy` documentation is acceptable as it is already documented). Charlie drafts a change and submits a pull request.

#### Story 3 (actual)

Rafael wanted to share a Kubernetes course from an online education provider. He submitted [PR #15962](https://github.com/kubernetes/website/pull/15962) to add the course to [Overview of Kubernetes Online Training](https://kubernetes.io/docs/tutorials/online-training/overview/). The PR was not approved because SIG Docs didn’t want to add a link to third-party content over which SIG Docs have no control.

#### Story 4 (actual)

> [PR #16203](https://github.com/kubernetes/website/pull/16203) removes Stackdriver and Elasticsearch vendor content. Since logging falls into the external add-ons category, SIG Docs decided to remove this vendor-specific content. That , which has not been meaningfully updated in three years. SIG Docs had buy-in from SIG Instrumentation Bugs for removal; however that PR was held pending the outcome of this KEP.

Reviewer 2: `. That , which` -> `, which`?

Reviewer: "Since logging falls into the external add-ons category, SIG Docs decided to remove this vendor-specific content that has not been meaningfully updated in three years."

Author: Updated

#### Story 5 (actual)

> In [PR #16766](https://github.com/kubernetes/website/pull/16766) @pouledodue proposed adding Hertzner Cloud Controller to the list of vendors that have implemented a cloud controller manager.

Reviewer 2: include the following here as well?
> however that PR was held pending the outcome of this KEP.

Author: Revised

## Summary

Feature docs are not a place for vendor pitches. Nevertheless, SIG Docs has seen pull requests that are more or less a request to place advertising content on the Kubernetes website.

Feature docs also contain content hosted in multiple places (“dual-sourced content”). A good practice for code project docs is to host single-sourced content only, and to provide links to other providers’ single-sourced content.
This KEP defines a policy on documentation content, so that authors can judge what is appropriate to propose and so that PR approvers can make consistent, fair decisions during the review process.

## Motivation

SIG Docs publishes Kubernetes documentation on kubernetes.io, in line with its [charter](https://github.com/kubernetes/community/blob/master/sig-docs/charter.md#scope), and also sets standards for website content. Prior to this KEP there were no clear guidelines or standards showing what documentation is acceptable.

The Kubernetes documentation is currently a mix of both 1) documentation describing the Kubernetes open source project; and 2) content describing how to install or use Kubernetes on several third party Kubernetes offerings.

> Some third party content is necessary in order for Kubernetes to function. For example, Docker, networking policy (CNI plugins), Ingress controllers, and logging all require third party components.

Reviewer 4: ```suggestion
Some third party content is necessary in order for Kubernetes to function. For example: Docker, networking policy (CNI plugins), Ingress controllers, and logging all require third party components.
```

Pages like [Logging Using Elasticsearch and Kibana](https://kubernetes.io/docs/tasks/debug-application-cluster/logging-elasticsearch-kibana/) are highly specific to a third party offering and seem more like third party product documentation than Kubernetes open source documentation.

> The goal of this KEP is to reach and document a consensus on what types of third party content are appropriate for inclusion in Kubernetes documentation; standards for including third-party content; and to drive a consistent policy toward third party content inclusion.

Reviewer 4: ```suggestion
The goal of this KEP is to reach and document a consensus on what types of third party content are appropriate for inclusion in Kubernetes documentation; standards for including third-party content; and to create consistent policies for docs handle third-party and dual-sourced content.
```

> This KEP focuses on the following issues related to third party content inclusion:

Reviewer 4: ```suggestion
This KEP focuses on the following issues:
```

1. What third party content is appropriate for inclusion in the Kubernetes documentation?
1. Does third party content in sections such as [Getting Started](https://kubernetes.io/docs/setup/) in the docs provide sufficient value to the reader that they should remain?
1. Is there a list of content pages that are so focused on third party product usage that they should be removed or updated from the Kubernetes documentation?
1. When should the Kubernetes documentation host third party documents that we are not the source/authority of?
1. How does the Kubernetes project handle third party content that is not kept up to date or hosts?
1. Can feature owners flag when third party content is *required*, as opposed to preferable or common?
1. Who decides when to include third-party content?
1. What standard of quality and review must be met before docs include third-party content?
1. To what extent should SIG Docs advocate for third-party content providers to host their own content, or decline to host third-party content altogether?

### Goals

> - Policy on including content about third party products in Kubernetes web documentation.

Reviewer 4: ```suggestion
- Create a policy for how to include or exclude third party content in Kubernetes documentation.
- Create a policy for how to handle dual-sourced content.
- Publish clear, transparent guidelines for contributors and reviewers on how to handle third-party and dual-sourced content.
```

#### Story 3 (actual)

Rafael wanted to share a Kubernetes course from an online education provider. He submitted [PR #15962](https://github.com/kubernetes/website/pull/15962) to add the course to [Overview of Kubernetes Online Training](https://kubernetes.io/docs/tutorials/online-training/overview/). The PR was not approved because SIG Docs didn’t want to add a link to third-party content over which SIG Docs have no control.

#### Story 4 (actual)

[PR #16203](https://github.com/kubernetes/website/pull/16203) removes Stackdriver and Elasticsearch vendor content. Since logging falls into the external add-ons category, SIG Docs decided to remove this vendor-specific content. That , which has not been meaningfully updated in three years. SIG Docs had buy-in from SIG Instrumentation Bugs for removal; however that PR was held pending the outcome of this KEP.

#### Story 5 (actual)

In [PR #16766](https://github.com/kubernetes/website/pull/16766) @pouledodue proposed adding Hertzner Cloud Controller to the list of vendors that have implemented a cloud controller manager.

### Implementation Details/Notes/Constraints

> SIG Docs approvers will need to consider intent. The agreed policy should be clear enough so that there are no or very few cases where intent needs to be considered.

Reviewer 4: ```suggestion
While SIG Docs approvers may occasionally need to consider intent, any policies produced by this KEP must be clear enough to minimize cases where intent needs to be considered.
```

1. When should the Kubernetes documentation host third party documents that we are not the source/authority of?
1. How does the Kubernetes project handle third party content that is not kept up to date or hosts?
1. Can feature owners flag when third party content is *required*, as opposed to preferable or common?
1. Who decides when to include third-party content?
1. What standard of quality and review must be met before docs include third-party content?
1. To what extent should SIG Docs advocate for third-party content providers to host their own content, or decline to host third-party content altogether?

### Goals

- Create a policy for how to include or exclude third party content in Kubernetes documentation.
- Create a policy for how to handle dual-sourced content.
- Publish clear, transparent guidelines for contributors and reviewers on how to handle third-party and dual-sourced content.
- Published guidelines for contributors and reviewers that clarify requirements for third party content and allow documentation reviewers to reject advertising.

### Non-Goals

- Outright removal of all content relating to vendors and projects outside the CNCF ecosystem.

## Proposal

> TBD. Detailed policies will come out of the discussion around this KEP.

Reviewer 4: ```suggestion
Revise the [content guide](https://github.com/kubernetes/website/blob/master/content/en/docs/contribute/style/content-guide.md#contributing-content) to address community concerns
```

Bob is a documentation approver. Bob explains that Alice’s proposed change does not meet community standards, because it is functionally an advertisement.

#### Story 2 (fictional)

Charlie uses Linux, specifically Ubuntu. Charlie notices that the page about installing `kubethingy` has instructions for installing `kubethingy` on Windows and on CentOS/RHEL but not on Ubuntu. Charlie reads the guidelines on content and sees that this kind of change is acceptable (Ubuntu is one of the most popular Linux distributions, and `kubethingy` documentation is acceptable as it is already documented). Charlie drafts a change and submits a pull request.

#### Story 3 (actual)

Rafael wanted to share a Kubernetes course from an online education provider. He submitted [PR #15962](https://github.com/kubernetes/website/pull/15962) to add the course to [Overview of Kubernetes Online Training](https://kubernetes.io/docs/tutorials/online-training/overview/). The PR was not approved because SIG Docs didn’t want to add a link to third-party content over which SIG Docs have no control.

#### Story 4 (actual)

> [PR #16203](https://github.com/kubernetes/website/pull/16203) removes Stackdriver and Elasticsearch vendor content. Since logging falls into the external add-ons category, SIG Docs decided to remove this vendor-specific content. That , which has not been meaningfully updated in three years. SIG Docs had buy-in from SIG Instrumentation Bugs for removal; however that PR was held pending the outcome of this KEP.

Author: ```suggestion
Website [PR #16203](https://github.com/kubernetes/website/pull/16203) removes Stackdriver and Elasticsearch vendor content. Since logging falls into the external add-ons category, SIG Docs decided to remove this vendor-specific content that had not been meaningfully updated in three years. SIG Docs had buy-in from SIG Instrumentation Bugs for removal; however that PR was held pending the outcome of this KEP.
```

#### Story 5 (actual)

> In [PR #16766](https://github.com/kubernetes/website/pull/16766) @pouledodue proposed adding Hertzner Cloud Controller to the list of vendors that have implemented a cloud controller manager.

Author: ```suggestion
In [PR #16766](https://github.com/kubernetes/website/pull/16766) @pouledodue proposed adding Hertzner Cloud Controller to the list of vendors that have implemented a cloud controller manager. That PR was held pending the outcome of this KEP.
```

> 

Reviewer 4: ```suggestion

#### Story 6 (actual)

As [hyperkube transitions to [third-party maintenance](https://github.com/kubernetes/kubeadm/issues/1889), it's unclear how to handle [hyperkube content in the Kubernetes docs](https://github.com/kubernetes/website/search?q=hyperkube&unscoped_q=hyperkube) or re-point related links.

```

 - [Auditing with Falco](https://kubernetes.io/docs/tasks/debug-application-cluster/falco/) (dual-sourced)
 - [Events in Stackdriver](https://kubernetes.io/docs/tasks/debug-application-cluster/events-stackdriver/) (vendor-specific content)
 - [Logging Using Elasticsearch and Kibana](https://kubernetes.io/docs/tasks/debug-application-cluster/logging-elasticsearch-kibana/) (vendor-specific content)
 - [Logging using Stackdriver](https://kubernetes.io/docs/tasks/debug-application-cluster/logging-stackdriver/) (vendor-specific content)

### Risks and Mitigations

- Rejecting a change that is ultimately deemed acceptable.
- Accepting a change that is ultimately judged to be unacceptable vendor promotion.
- It may not always be possible to achieve consensus on whether specific third-party content should be included.
  This may lead to resentment/resistance to future third party content.
  *Mitigation*: consider including folks that also work outside the relevant SIG or subproject.

## Design Details

### Graduation Criteria

**Note:** *this KEP does not target any release*

TBD

## Drawbacks

_Why should this KEP _not_ be implemented._

> TBD

Author: Thoughts on this, anyone?

Reviewer 5: I don't mean this to be snarky at all, but SIG-Docs currently makes the call as to what goes into the docs and what stays out of the docs. Arguably, it's in SIG-Docs's charter to set these standards themselves without the need of a KEP.

Reviewer 5: I think these policies should be transparent, regardless of whether they live in a kep or in the SIG-Docs charter.

Author: Thanks for the feedback. This KEP is at least partly there to document the _opportunity_ for discussion and to make things visible to people outside of SIG Docs.

  - TBD
editor: TBD
creation-date: 2019-10-20
last-updated: 2019-10-20
status: provisional
---

# doc-policies-for-third-party-content

## Table of Contents

<!-- toc -->
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [User Stories](#user-stories)
    - [Story 1 (fictional)](#story-1-fictional)
    - [Story 2 (fictional)](#story-2-fictional)
    - [Story 3 (actual)](#story-3-actual)
    - [Story 4 (actual)](#story-4-actual)
    - [Story 5 (actual)](#story-5-actual)
    - [Story 6 (actual)](#story-6-actual)
  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Design Details](#design-details)
  - [Graduation Criteria](#graduation-criteria)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
<!-- /toc -->

**Note:** This KEP does not target any release; SIG Docs follows a continuous release process for website content.

## Summary

This KEP seeks consensus on how Kubernetes docs handle two types of content:
1. Content from third-party providers
2. Content hosted on multiple sites ("dual-sourced content")

> Feature docs are not a place for vendor pitches. Nevertheless, SIG Docs receives pull requests to place advertising content on the Kubernetes website. Some PRs clearly do not belong in feature docs, but other instances are less clear.

Reviewer 5: Might be good to start this paragraph with the current goals of Kubernetes docs, something like "Kubernetes documentation seeks to teach Kubernetes users about how kubernetes works, how to user in-tree kubernetes features, and how to build on top of Kubernetes infrastucture."

Reviewer 5: Also, I suggest "technical documentation" not "feature documentation". The documentation is more than just the features of Kubernetes.

Feature docs also contain dual-sourced content. A good practice for code project docs is to host single-sourced content only, and to provide links to other providers’ single-sourced content. This simplifies version management and reduces the work required to maintain content.

> This KEP defines a policy on documentation content, so that authors can judge what is appropriate to propose and so that PR approvers can make consistent, fair decisions during the review process.

Reviewer 5: It seems like SIG-Docs is already judging what PRs are "legitimate" and what PRs aren't. This KEP seeks to make that process transparent to the community.

### Implementation Details/Notes/Constraints

While SIG Docs approvers may occasionally need to consider intent, any policies produced by this KEP must be clear enough to minimize cases where intent needs to be considered.

(At the time of writing) Kubernetes requires external software to implement Cluster Networking, Ingress, Persistent Storage, and Logging. Hyperlinking to vendor software and documentation _is_ allowed; creating “how to use” content is not.

Examples of allowed content:
 - Cluster Networking
   - https://kubernetes.io/docs/concepts/cluster-administration/networking/
   - https://kubernetes.io/docs/concepts/cluster-administration/addons/
 - Ingress Controllers
   - https://kubernetes.io/docs/concepts/services-networking/ingress-controllers/#additional-controllers
 - Persistent Volumes
   - https://kubernetes.io/docs/concepts/storage/persistent-volumes/#expanding-persistent-volumes-claims

Examples of content that would not be allowed:

>  - [Install a Network Policy Provider](https://kubernetes.io/docs/tasks/administer-cluster/network-policy-provider/) and child pages: how to use Calico, Cilium, Kube-router, Romana, and Weave Net for NetworkPolicy

Reviewer 5: Network policy is an interesting one. I can understand pushing back on this particular doc which describes a specific third-party implementation, but would we keep a doc like this page with it's list of supported providers? https://kubernetes.io/docs/tasks/administer-cluster/declare-network-policy/

I think there's some grey area here.

Reviewer 6: Agree - there is no "built-in" implementation.  But there's an implication that things listed here are "real" or "supported by the project", which is not necessarily true.  It also becomes an objective for new implementation - get listed or remain unknown.

project docs is to host single-sourced content only, and to provide
links to other providers’ single-sourced content. This simplifies
version management and reduces the work required to maintain content.

This KEP proposes that Kubernetes documentation sets these boundaries:

- Kubernetes documentation includes 

This KEP defines how to handle third-party and dual-sourced content in 
documentation, so that authors can
judge what is appropriate to propose and so that PR approvers can make
consistent, fair decisions during the review process.

## Motivation

SIG Docs publishes Kubernetes
documentation on kubernetes.io in line with its
[charter](https://github.com/kubernetes/community/blob/master/sig-docs/charter.md#scope)
and sets standards for website content. Prior to this KEP, there are no
clear guidelines or standards for third-party and dual-sourced content.

The Kubernetes documentation is currently a mix of both 1) documentation
describing the Kubernetes open source project; and 2) content describing
how to install or use Kubernetes on several third party Kubernetes
offerings.

Some third party content is necessary in order for Kubernetes to

> function. For example: Docker, networking policy (CNI plugins), Ingress

Reviewer 7: minor nit:  swap Docker for container runtimes (containerd, CRI-o, docker, etc)

      - [Story 3 (actual)](#story-3-actual)
      - [Story 4 (actual)](#story-4-actual)
      - [Story 5 (actual)](#story-5-actual)
      - [Story 6 (actual)](#story-6-actual)
    - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
    - [Risks and Mitigations](#risks-and-mitigations)
  - [Design Details](#design-details)
    - [Graduation Criteria](#graduation-criteria)
- [Drawbacks](#drawbacks)
  - [Alternatives](#alternatives)
<!-- /toc -->

**Note:** This KEP does not target any release; SIG Docs follows a continuous
release process for website content.

## Summary

This KEP seeks consensus on how Kubernetes docs handle two types of content:

1. Content from or about third-party providers ("third-party content")

Minimize and eliminate third-party content except when necessary for Kubernetes 
to function in-project.

2. Content hosted on multiple sites ("dual-sourced content")

Minimize and eliminate dual-sourced content except when necessary for Kubernetes
to function in-project.

**Note:** This KEP defines "in project" to mean projects in the Kubernetes org, 
which includes the [kubernetes](https://github.com/kubernetes) and 

> [kubernetes-sigs](https://github.com/kubernetes) repositories.

Reviewer 8: [kubernetes-sigs](https://github.com/kubernetes-sigs/)

> [kubernetes-sigs](https://github.com/kubernetes) repositories.

Author: ```suggestion
[kubernetes-sigs](https://github.com/kubernetes-sigs) repositories.
```

## Introduction

Kubernetes documentation teaches Kubernetes users about how
Kubernetes works, how to use in-project Kubernetes features, and how to
build on top of Kubernetes infrastucture.

Feature docs are not a place for vendor pitches. Nonetheless, SIG Docs sometimes
receives pull requests to place advertising-like content on the Kubernetes
website. Some PRs clearly do not belong in feature docs, but other
instances are less clear.

Feature docs also contain dual-sourced content. A good practice for code
project docs is to host single-sourced content only, and to provide
links to other providers’ single-sourced content. This simplifies
version management and reduces the work required to maintain content.

This KEP proposes that Kubernetes documentation sets these boundaries:

> - Kubernetes documentation includes 

Reviewer 9: I don't understand this line... is this a fragment or am I misreading it?

Reviewer 4: Nope, it's a fragment. I've removed it.

This KEP defines how to handle third-party and dual-sourced content in 
documentation, so that authors can
judge what is appropriate to propose and so that PR approvers can make
consistent, fair decisions during the review process.

## Motivation

SIG Docs publishes Kubernetes
documentation on kubernetes.io in line with its
[charter](https://github.com/kubernetes/community/blob/master/sig-docs/charter.md#scope)
and sets standards for website content. Prior to this KEP, there are no
clear guidelines or standards for third-party and dual-sourced content.

The Kubernetes documentation is currently a mix of both 1) documentation
describing the Kubernetes open source project; and 2) content describing
how to install or use Kubernetes on several third party Kubernetes
offerings.

Some third party content is necessary in order for Kubernetes to
function. For example: container runtimes (containerd, CRI-o, Docker), 
networking policy (CNI plugins), Ingress controllers, and logging all require 

> third party components. Pages like [Logging Using Elasticsearch and Kibana](https://kubernetes.io/docs/tasks/debug-application-cluster/logging-elasticsearch-kibana/) 

Reviewer 4: @thockin 
> So what is the principle we want?

We have language that talks about [linking versus hosting](https://github.com/kubernetes/enhancements/pull/1327/files#diff-7c573870e6b3f25f38e3837aaaa193dcR206-R207) in order to avoid exactly this scenario, where _no one_ knows whether a doc is fresh. However, linking is no guarantee of freshness: links break, redirect unhelpfully, and go stale too. We need to future proof and stale proof our content as best we're able. Drawing boundaries at third party content seems like the best way to do that.

> third party components. Pages like [Logging Using Elasticsearch and Kibana](https://kubernetes.io/docs/tasks/debug-application-cluster/logging-elasticsearch-kibana/) 

Reviewer 10: The principle here is that third party docs should only be included where absolutely necessary. The E+K doc says it is specific to GCE and is out of date from what I can tell (and therefore would be eligible for removal unless it was updated, or better linked externally). If we are only using this as an example to highlight the principle though, maybe something more black and white like [Logging Using Stackdriver](https://kubernetes.io/docs/tasks/debug-application-cluster/logging-stackdriver/) would be better. Again, this doesn't change the principle or intent, but is a more clear example without having to read the linked document. Swapping the link would be sufficient.

are highly specific to a third party offering and seem more like third party 
product documentation than Kubernetes open source documentation.

### Goals

The goal of Kubernetes documentation is to accurately document in-project
functionality for Kubernetes, and to eliminate barriers to effective
contribution and understanding.

The goal of this KEP is to reach and document a consensus on what
types of third-party content are appropriate for inclusion in Kubernetes
documentation; standards for including third-party content; and to create
consistent policies for docs handle third-party and dual-sourced content.

To address its goal, this KEP focuses on the following issues:

<del>

1. What third party content is appropriate for inclusion in the Kubernetes
documentation?

> Proposed: Third-party content is permitted if it's required for Kubernetes to

Reviewer 8: it's -> it is ?

Reviewer 4: "It's" is grammatically correct here; is it just a preference to avoid contractions?

function in-project.

1. Does third party content in sections such as [Getting Started](https://kubernetes.io/docs/setup/)
in the docs provide sufficient value to the reader that they should remain?

Casual consensus says yes, with one modification:
- Eliminate the [production environment table](https://kubernetes.io/docs/setup/#production-environment)
with a link to [certified conformance partners](https://kubernetes.io/partners/#conformance).

1. Is there a list of content pages that are so focused on third party product
usage that they should be removed or updated from the Kubernetes documentation?

See https://github.com/kubernetes/website/issues/15748.

1. When should the Kubernetes documentation host third party content that isn't
maintained by a Kubernetes SIG?

As infrequently as possible, with linking preferred to hosting. 

1. How does the Kubernetes project handle third party content that is not kept
up to date or hosts?

> If content isn't refreshed within 180 days, notify stakeholders of 90 days to 

Reviewer 8: Can we please explicitly say that we will do this "notify" step to all existing content that is on the chopping block?

their discretion, preferably by linking to the third party's own documentation.

1. What standard of quality and review must be met before docs include
third-party content?

Third-party content must be necessary for Kubernetes to function in-project.

1. To what extent should SIG Docs advocate for third-party content providers to 
host their own content, or decline to host third-party content altogether?

Kubernetes docs publish third-party content only if:

- It's necessary for Kubernetes to function. For example: container runtimes 
(containerd, CRI-o, Docker), networking policy (CNI plugins), Ingress 
controllers, and logging.

- It's an applied example of another project in the Kubernetes GitHub org. This
includes the [kubernetes](https://github.com/kubernetes) and 
[kubernetes-sigs](https://github.com/kubernetes-sigs) repositories.

Third-party content should be linked instead of hosted whenever possible. 

</del>

1. Clearly define what documentation is required so that readers understand
   how to deploy, operate and consume Kubernetes clusters using features from
   in-project code and its mandatory dependencies.

### Non-Goals

> 1. Outright removal of all content relating to vendors and projects

Reviewer 8: The doc upto this point did not mention anything about `CNCF` it seems jarring that we refer to `CNCF` all of a sudden here ... can we switch to `in-project` language here?

   outside the CNCF ecosystem.

## Proposal

1. Revise the [content guide](https://github.com/kubernetes/website/blob/master/content/en/docs/contribute/style/content-guide.md#contributing-content) to achieve the KEP goal:

- Specify that Kubernetes docs are limited to content required for Kubernetes to

> function in-project. Docs may include third-party content for components that 

Reviewer 8: Does `third-party` include open source projects AND vendor specific? or just vendor specific? A definition like the `in-project` would be useful here

Reviewer 4: @dims 
> Does third-party include open source projects AND vendor specific? or just vendor specific?

Ideally we'd be leaving this to the discretion of the relevant SIGs. In practice, allowing vendor-specific content created the race for inclusion we see now. 

Specifically, I don't want to deal with more angry Slack messages from vendors who can't get a mention in the same page as their competitor because their competitor has a contributor in the SIG and they don't.

Unless you have a better solution, let's limit third-party content to OSS projects that are necessary to run Kubernetes. To stem off inevitable what-aboutism re:Docker, can we acknowledge that Docker is an edge case without having to specifically say that Docker is the sole exception?

require a third-party solution to function. Docs may include content for 
other projects in the Kubernetes org. Third-party content must be linked 
whenever possible, rather than duplicated or hosted in k/website.

2. Revise the documentation when the KEP is approved:

> - **Third-party content:** Notify stakeholders via GitHub issues in k/website

Reviewer 8: Can you also please bunch up these removals and post to kubernetes-sig-docs mailing list as well? (since github notifications are busted for a bunch of folks)?

that third-party content will be removed after 90 days if content isn't required
for Kubernetes to function in-project.

This limits the impact to out-of-project content and gives current stakeholders 
approximately one Kubernetes release cycle to migrate 
third-party content to an alternate platform before removing content from 
Kubernetes docs.

- **Dual-sourced content:** Where sourcing is obvious, replace dual-sourced
content with links to an authoritative single source. Where sourcing is unclear,

> notify stakeholders via GitHub issues in k/website that dual-sourced content

Reviewer 8: sentence seems incomplete?

Reviewer 4: Fixed.

In all cases where content would be removed, provide adequate time for the

> relevant SIG to review changes in content.

Reviewer 8: "relevant SIG to review changes and notify stake holders" right?

SIG Docs may add its own guidelines for writing and reviewing ambiguous content.

For example:
> Kubernetes requires out-of-tree software and tools to implement: cluster
> networking, Ingress, persistent storage, and logging. Hyperlinking to vendor software
> and documentation _is_ allowed; creating “how to use” content for a specific vendor
> is discouraged.

Pages that fit with that example guideline:
 - Cluster Networking
   - https://kubernetes.io/docs/concepts/cluster-administration/networking/
   - https://kubernetes.io/docs/concepts/cluster-administration/addons/
 - Ingress Controllers
   - https://kubernetes.io/docs/concepts/services-networking/ingress-controllers/#additional-controllers
 - Persistent Volumes
   - https://kubernetes.io/docs/concepts/storage/persistent-volumes/#expanding-persistent-volumes-claims

Pages to review and possibly revise, if that guideline were in place:
 - [Install a Network Policy Provider](https://kubernetes.io/docs/tasks/administer-cluster/network-policy-provider/) and child pages: how to use Calico, Cilium, Kube-router, Romana, and Weave Net for NetworkPolicy
 - [Audit](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/)

>  - [Use fluentd to collect and distribute audit events from log file](https://kubernetes.io/docs/tasks/debug-application-cluster/audit/#use-fluentd-to-collect-and-distribute-audit-events-from-log-file) (dual-sourced)

Reviewer 11: This is an interesting one from my perspective. 

It's the only other project on this list officially under the CNCF hood. All other CNCF projects explicitly call out and document their compatibility with Kubernetes. For example, [Vitess](https://vitess.io/) calls out "Run Vitess on Kubernetes" as a top-level menu item in docs. Moreover, there are other CNCF projects like [etcd](https://etcd.io/) which are almost required for running a cluster.

So I think there's two questions we need to think about. One is "What do we do with _third party_ content?" and the second is "Do other CNCF projects count as _third party_?"

Author: BTW, Falco is CNCF incubating. It seems to me like it's in a similar position.

Author: > "Do other CNCF projects count as third party?"

That seems worth answering. The PR as it stands sees things outside Kubernetes as third party. That's definitely open to revision.

Reviewer 4: @celestehorgan
> "What do we do with third party content?"

Remove it.

> "Do other CNCF projects count as third party?"

They count as third-party (and therefore removable) if they're not necessary for Kubernetes to run in project.

Reviewer 4: @sftim 
> The PR as it stands sees things outside Kubernetes as third party. That's definitely open to revision.

Respectfully, I'd like to limit further revision to this KEP's approvers only. Five months is more than enough time for review. In the case of how we define content, the current language is [two months old](https://github.com/kubernetes/enhancements/pull/1327#issuecomment-579546667), which also seems adequate.
